### DISCUSSION TITLE
Name for a string interpolation operation

### DISCUSSION THREAD

**Discussion by fmontesi** [jolie/jolie#426] [Polls]
I'd like some feedback on an operation for string interpolation that I've just provisionally added to the `StringUtils` library, based on the Apache commons text library. It can be used like this:
```jolie
// su is an instance of StringUtils
println@console( ip@su( "Hello ${name}" { name = "Homer" } ) )
// prints "Hello Homer"
```

`ip` stands for "interpolate". I could call it "interpolate" or "substitute", but I was concerned it'd be too long since this operation is likely to be used quite a bit. However, it'd be clearer. I could also call it "render" or "eval", but I'm concerned they're too general.

**Comment by jclark** [comment]
subst

**Reply by fmontesi** [reply]
Really tempting. :)

**Comment by mperessotti** [comment]
I think that the terminology "interpolation" does not apply here since values for the placeholders in the string literal need to be passed as arguments to the operation.  In string interpolation, placeholders are variable names (or in some languages arbitrary expressions) and they are evaluated in the current context.

**Reply by fmontesi** [reply]
Right... maybe `subst` by @jclark or the `fmt`/`format` examples given by Alexander in the chat might be better then.

**Reply by fmontesi** [reply]
Btw I was thinking (wrongly) about interpolation because this is an experiment-as-a-library to reach interpolation later on. But one thing at a time. :)

**Comment by Nissen96** [comment]
I would suggest `format` or `fmt` as in languages like Python.
In Python it's used in much the same way with braces (either named like in your example, or empty for positional values). Examples:
```py
"{} {}! How are you?".format("Hello", "Fabrizio")
"{greeting} {name}! How are you?".format(name="Jolie", greeting="Hi")
```
It also has a range of modifiers, which you could imagine would be useful to add to Jolie as well, e.g.
```py
# Format as float with 2 decimals
>>> "The average was {avg:.2f}".format(avg=78.234876)
'The average was 78.23'

# Print an integer in binary (similar modifiers exist for octal and hex)
>>> "{n} in binary is {n:b}".format(n=42)
'42 in binary is 101010'
 
 # Left- or right-padding with spaces or zeros (e.g. for formatting tables)
>>> "{n} left-padded to length 8 with zeros is {n:010}".format(n=1234)
'1234 left-padded to length 8 with zeros is 0000001234'
```
`subst` would work perfectly fine as well, but if you intend to at some point add more formatting options like above, I would use `format` or `fmt`.

**Reply by fmontesi** [reply]
`fmt` looks neat. Even `format` is short enough to be added as is maybe.

**Comment by thesave** [comment]
Since what the operation does seems akin to template rendering (where you give the operation a template and a data structure for the renderer to generate the output string) `render` seems appropriate (à la Liquid et al. [1-5], just to mention some examples).

[1] https://github.com/Shopify/liquid#how-to-use-liquid
[2] https://aui.github.io/art-template/docs/#Render-template
[3] https://jinja.palletsprojects.com/en/3.1.x/api/#basics
[4] https://twig.symfony.com/doc/3.x/api.html#rendering-templates
[5] https://github.com/lets-blade/blade#render-template

**Reply by fmontesi** [reply]
That was the first name I tried. But then I thought that maybe people have different expectations from a templating system, like being able to include other templates. Here including a string within a string would not render recursively.

**Comment by fmontesi** [comment]
Thank you for your input, everyone! :)

I've pushed a change that renames the operation to `fmt`, and introduced the possibility of doing some formatting, e.g.,
```jolie
fmt@su( "Up to {pct,number,percent}" { pct = 0.6 } ) // "Up to 60%"
fmt@su( "The disk contains {files,number,integer} file(s)" { files = 12312332 } ) // "The disk contains 12,312,332 file(s)"
```

It's based on the Java `MessageFormat` library, but it's a bit more modern-looking and parameters are named. ;-)

Comments/more suggestions are of course welcome!
